src/pojo-interfaces.ts

```diff
@@ -0,0 +1,418 @@
+/* eslint-disable license-header/header */
+/**
+ * This is just a strawman proposal;
+ * the actual new interfaces will not be implemented through this file.
+ *
+ * They're also not set in stone. Actual implementation experience might result in further tweaks.
+ * Names can still be changed if we propose it as an actual specification, of course.
+ *
+ * The design goals of these interfaces are:
+ * - JSON.parse(JSON.stringify(data)) is lossless
+ * - Optimised for reading and copying in an expected shape in stateful (Solid) applications
+ * - fromRdfJsDataset(toRdfJsDataset(data)) is lossless
+ *
+ * Non-goals:
+ * - Optimising for streaming data
+ * - Optimising for unknown use cases
+ * - Be perfect from the get-go
+ * - Be immediately useful outside of solid-client
+ * @module
+ */
+
+import { dataset as rdfJsDataset } from "@rdfjs/dataset";
+import RdfJsDataFactory from "@rdfjs/data-model";
+import * as RdfJs from "rdf-js";
+import { IriString } from "./interfaces";
+import { xmlSchemaTypes } from "./datatypes";
+
+const localNodeSkolemPrefix = "https://inrupt.com/.well-known/sdk-local-node/" as const;
+type LocalNodeIri = `${typeof localNodeSkolemPrefix}${string}`;
+export type LocalNodeName = string;
+export type Objects = Readonly<
+  Partial<{
+    literals: Readonly<Record<IriString, readonly string[]>>;
+    langStrings: Readonly<Record<string, readonly string[]>>;
+    namedNodes: ReadonlyArray<LocalNodeIri | IriString>;
+    // By abstracting over the specific blank nodes, we can neatly skip over the
+    // issue of fetching the same data twice and then not being able to reconcile
+    // different instances of the same blank nodes.
+    blankNodes: readonly Predicates[];
+    // TODO: Do we need RDF/JS Variables as well, or are they just for SPARQL?
+  }>
+>;
+
+export type Predicates = Readonly<Record<IriString, Objects>>;
+
+export type Subject = Readonly<{
+  type: "Subject";
+  url: LocalNodeIri | IriString;
+  predicates: Predicates;
+}>;
+
+export type Graph = Readonly<Record<IriString, Subject>>;
+
+export type Dataset = Readonly<{
+  type: "Dataset";
+  graphs: Readonly<Record<"default" | IriString, Graph>>;
+}>;
+
+/**
+ * Runtime freezing might be too much overhead;
+ * if so, this function allows us to replace it by a function
+ * that merely marks its input as Readonly<> for static analysis.
+ */
+const freeze: typeof Object.freeze = Object.freeze;
+
+export function fromRdfJsDataset(rdfJsDataset: RdfJs.DatasetCore): Dataset {
+  const dataset: Dataset = {
+    graphs: {},
+    type: "Dataset",
+  };
+
+  const quads = Array.from(rdfJsDataset);
+  const quadsWithoutBlankNodes = quads.filter(
+    (quad) => !isBlankNode(quad.subject)
```

Comment:
Are we deliberately only filtering on Subject BNodes? i.e., should we also be filtering on Object BNodes too...?
Probably warrants a comment regardless (i.e., why are we filtering out quads with BNodes at all?), and how do these quads get added back in (i.e., I assume they do get added back in!)...?

Reply:
b7714c88f200cc1cf61659f185125a49fb4490ca